Swap `isFirstInComplexSelector` and `isLastInComplexSelector`

nt1m opened the pull request:
#### 8f719fa47034a9e9bc8f572ef4f9b4ce443c61ad
<pre>
Swap `isFirstInComplexSelector` and `isLastInComplexSelector`
<a href="https://bugs.webkit.org/show_bug.cgi?id=298173">https://bugs.webkit.org/show_bug.cgi?id=298173</a>
<a href="https://rdar.apple.com/159557880">rdar://159557880</a>

Reviewed by Antti Koivisto.

The meaning is reversed since the rename from `tagHistory` to `inComplexSelector` in 299193@main.

`tagHistory` uses memory order, `inComplexSelector` uses logical order which is reversed from the memory order.

Also rename `CSSSelector::firstInCompound` to `CSSSelector::lastInCompound` to reflect what it actually represents.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::lastInCompound const):
(WebCore::CSSSelector::firstInCompound const):
* Source/WebCore/css/CSSSelector.h:
(WebCore::CSSSelector::precedingInComplexSelector const):
* Source/WebCore/css/CSSSelectorList.cpp:
(WebCore::CSSSelectorList::CSSSelectorList):
(WebCore::CSSSelectorList::listSize const):
* Source/WebCore/css/CSSSelectorList.h:
(WebCore::CSSSelectorList::const_iterator::operator++):
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::constructFragmentsInternal):
* Source/WebCore/dom/SelectorQuery.cpp:
(WebCore::isSingleTagNameSelector):
(WebCore::isSingleClassNameSelector):
(WebCore::isSingleAttributeExactSelector):
(WebCore::SelectorDataList::SelectorDataList):
* Source/WebCore/style/RuleFeature.cpp:
(WebCore::Style::makePseudoClassInvalidationKey):

Canonical link: <a href="https://commits.webkit.org/299397@main">https://commits.webkit.org/299397@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a5f9592d8446488eabd319a70049dd4ae858ce75

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118888 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38569 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29220 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125081 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70951 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a01047d9-3108-414d-a6d2-6f9a70c8dd31) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120766 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39265 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47151 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/90224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/adb9de3e-08be-4818-83d1-bb921caf9755) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121841 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/31296 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/106589 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/70731 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e4333add-78dc-4ad9-ba13-08035781bad3) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/30359 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/24703 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68740 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/100739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/24890 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128128 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45795 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/34589 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/98885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46161 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/102809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/98665 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/44113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/22120 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/42360 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18938 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45665 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51343 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45131 "Built successfully") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/48475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46815 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/css/CSSSelectorList.cpp:
```diff
@@ -60,28 +60,28 @@ CSSSelectorList::CSSSelectorList(MutableCSSSelectorList&& selectorVector)
             ASSERT(!m_selectorArray[arrayIndex].isLastInSelectorList() || (flattenedSize == arrayIndex + 1));
             if (current)
-                m_selectorArray[arrayIndex].m_isLastInComplexSelector = false;
+                m_selectorArray[arrayIndex].m_isFirstInComplexSelector = false;
```
You should add a comment near these bits that they are first/last in the logical selector order, not the memory order.

nt1m committed:
Swap `isFirstInComplexSelector` and `isLastInComplexSelector`
https://bugs.webkit.org/show_bug.cgi?id=298173
rdar://159557880

Reviewed by Antti Koivisto.

The meaning is reversed since the rename from `tagHistory` to `inComplexSelector` in 299193@main.

`tagHistory` uses memory order, `inComplexSelector` uses logical order which is reversed from the memory order.

Also rename `CSSSelector::firstInCompound` to `CSSSelector::lastInCompound` to reflect what it actually represents.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::lastInCompound const):
(WebCore::CSSSelector::firstInCompound const):
* Source/WebCore/css/CSSSelector.h:
(WebCore::CSSSelector::precedingInComplexSelector const):
* Source/WebCore/css/CSSSelectorList.cpp:
(WebCore::CSSSelectorList::CSSSelectorList):
(WebCore::CSSSelectorList::listSize const):
* Source/WebCore/css/CSSSelectorList.h:
(WebCore::CSSSelectorList::const_iterator::operator++):
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::constructFragmentsInternal):
* Source/WebCore/dom/SelectorQuery.cpp:
(WebCore::isSingleTagNameSelector):
(WebCore::isSingleClassNameSelector):
(WebCore::isSingleAttributeExactSelector):
(WebCore::SelectorDataList::SelectorDataList):
* Source/WebCore/style/RuleFeature.cpp:
(WebCore::Style::makePseudoClassInvalidationKey):

Canonical link: https://commits.webkit.org/299397@main
